pkgs/development/beam-modules/build-mix.nix

```diff
@@ -1,100 +1,90 @@
-{ stdenv, writeText, elixir, erlang, hex, lib }:
+{ stdenv, lib, elixir, erlang, hex, rebar, rebar3, fetchMixDeps, makeWrapper, git }:
 
 { name
 , version
 , src
-, setupHook ? null
-, buildInputs ? []
-, beamDeps ? []
-, postPatch ? ""
-, compilePorts ? false
-, installPhase ? null
-, buildPhase ? null
-, configurePhase ? null
-, meta ? {}
+, nativeBuildInputs ? [ ]
+, meta ? { }
 , enableDebugInfo ? false
-, ... }@attrs:
-
-with lib;
-
+, mixEnv ? "prod"
+, compileFlags ? [ ]
+, defaultEnvVars ? true
+, mixDeps ? null
+, ...
+}@attrs:
 let
+  overridable = builtins.removeAttrs attrs [ "compileFlags" ];
 
-  debugInfoFlag = lib.optionalString (enableDebugInfo || elixir.debugInfo) "--debug-info";
+in
+stdenv.mkDerivation (overridable // {
+  name = "${name}-${version}";
+  inherit version;
 
-  shell = drv: stdenv.mkDerivation {
-          name = "interactive-shell-${drv.name}";
-          buildInputs = [ drv ];
-    };
+  dontStrip = true;
 
-  bootstrapper = ./mix-bootstrap;
+  inherit src;
 
-  pkg = self: stdenv.mkDerivation ( attrs // {
-    name = "${name}-${version}";
-    inherit version;
+  nativeBuildInputs = nativeBuildInputs ++ [ erlang hex elixir makeWrapper git ];
 
-    dontStrip = true;
+  MIX_ENV = mixEnv;
+  MIX_DEBUG = if enableDebugInfo then 1 else 0;
+  HEX_OFFLINE = 1;
+  DEBUG = if enableDebugInfo then 1 else 0; # for Rebar3 compilation
+  # the api with `mix local.rebar rebar path` makes a copy of the binary
+  MIX_REBAR = "${rebar}/bin/rebar";
+  MIX_REBAR3 = "${rebar3}/bin/rebar3";
 
-    inherit src;
+  postUnpack = ''
+    export HEX_HOME="$TEMPDIR/hex"
+    export MIX_HOME="$TEMPDIR/mix"
+    # compilation of the dependencies will require
+    # that the dependency path is writable
+    # thus a copy to the TEMPDIR is inevitable here
+    export MIX_DEPS_PATH="$TEMPDIR/deps"
 
-    setupHook = if setupHook == null
-    then writeText "setupHook.sh" ''
-       addToSearchPath ERL_LIBS "$1/lib/erlang/lib"
-    ''
-    else setupHook;
+    # Rebar
+    export REBAR_GLOBAL_CONFIG_DIR="$TEMPDIR/rebar3"
+    export REBAR_CACHE_DIR="$TEMPDIR/rebar3.cache"
 
-    inherit buildInputs;
-    propagatedBuildInputs = [ hex elixir ] ++ beamDeps;
+    cp --no-preserve=mode -R "${mixDeps}" "$MIX_DEPS_PATH"
```

Comment:
It is possible to have some Elixir apps that have no dependencies, to me it would feel weird to ask for a `fetchMixDeps` of no deps. I would personnaly go for something like this:

```nix
''
${optionalString (mixDeps != null)
  "cp ... ${mixDeps} ..."}"
''
```